apps/src/JSInterpreter.js

```diff
@@ -40,16 +42,45 @@ var JSInterpreter = module.exports = function (options) {
     options.code += '\nwhile (true) { var obj = getCallback(); ' +
       'if (obj) { var ret = obj.fn.apply(null, obj.arguments ? obj.arguments : null);' +
                  'setCallbackRetVal(ret); }}';
+
+    codegen.createNativeFunctionFromInterpreterFunction = function (intFunc) {
+      return function () {
+        if (self.initialized()) {
+          self.queueEvent(intFunc, arguments);
+          
+          if (self.executeLoopDepth === 0) {
+            // Execute the interpreter and if a return value is sent back from the
+            // interpreter's event handler, pass that back in the native world
+
+            // NOTE: the interpreter will not execute forever, if the event handler
+            // takes too long, executeInterpreter() will return and the native side
+            // will just see 'undefined' as the return value. The rest of the interpreter
+            // event handler will run in the next onTick(), but the return value will
+            // no longer have any effect.
+            self.executeInterpreter(false, true);
+            return self.lastCallbackRetVal;
```

Comment:
This is code that was moved, so the code and comment are identical. When `executeInterpreter` is called with 2nd parameter (`runUntilCallbackReturn`) set to `true`, the `lastCallbackRetVal` is deleted (set back to `undefined`). So it will always be either `undefined` or the value stored by the callback.